server/src-lib/Hasura/Eventing/ScheduledTrigger.hs

```diff
@@ -0,0 +1,435 @@
+      let decodedHeaders = map (decodeHeader logEnv stiHeaders) headers
+      either
+        (processError pgpool se decodedHeaders)
+        (processSuccess pgpool se decodedHeaders)
+        res
+
+processError
+  :: (MonadIO m, MonadError QErr m)
+  => Q.PGPool -> ScheduledEventFull -> [HeaderConf] -> HTTPErr a -> m ()
+processError pgpool se decodedHeaders err = do
+  let invocation = case err of
+        HClient excp -> do
+          let errMsg = TBS.fromLBS $ J.encode $ show excp
+          mkInvocation se 1000 decodedHeaders errMsg []
+        HParse _ detail -> do
+          let errMsg = TBS.fromLBS $ J.encode detail
+          mkInvocation se 1001 decodedHeaders errMsg []
+        HStatus errResp -> do
+          let respPayload = hrsBody errResp
+              respHeaders = hrsHeaders errResp
+              respStatus = hrsStatus errResp
+          mkInvocation se respStatus decodedHeaders respPayload respHeaders
+        HOther detail -> do
+          let errMsg = (TBS.fromLBS $ J.encode detail)
+          mkInvocation se 500 decodedHeaders errMsg []
+  liftExceptTIO $
+    Q.runTx pgpool (Q.RepeatableRead, Just Q.ReadWrite) $ do
+    insertInvocation invocation
+    retryOrMarkError se err
+
+retryOrMarkError :: ScheduledEventFull -> HTTPErr a -> Q.TxE QErr ()
+retryOrMarkError se@ScheduledEventFull {..} err = do
+  let mRetryHeader = getRetryAfterHeaderFromHTTPErr err
+      mRetryHeaderSeconds = parseRetryHeaderValue =<< mRetryHeader
+      triesExhausted = sefTries >= rcstNumRetries sefRetryConf
+      noRetryHeader = isNothing mRetryHeaderSeconds
+  if triesExhausted && noRetryHeader
+    then do
+      markError
+    else do
+      currentTime <- liftIO getCurrentTime
+      let delay = fromMaybe (round $ rcstIntervalSec sefRetryConf) mRetryHeaderSeconds
+          diff = fromIntegral delay
+          retryTime = addUTCTime diff currentTime
+      setRetry se retryTime
+  where
+    markError =
+      Q.unitQE
+        defaultTxErrorHandler
+        [Q.sql|
+        UPDATE hdb_catalog.hdb_scheduled_events
+        SET error = 't', locked = 'f'
+        WHERE id = $1
+      |] (Identity sefId) True
+
+processSuccess
+  :: (MonadIO m, MonadError QErr m)
+  => Q.PGPool -> ScheduledEventFull -> [HeaderConf] -> HTTPResp a -> m ()
+processSuccess pgpool se decodedHeaders resp = do
+  let respBody = hrsBody resp
+      respHeaders = hrsHeaders resp
+      respStatus = hrsStatus resp
+      invocation = mkInvocation se respStatus decodedHeaders respBody respHeaders
+  liftExceptTIO $
+    Q.runTx pgpool (Q.RepeatableRead, Just Q.ReadWrite) $ do
+    insertInvocation invocation
+    markSuccess
+  where
+    markSuccess =
+      Q.unitQE
+        defaultTxErrorHandler
+        [Q.sql|
+        UPDATE hdb_catalog.hdb_scheduled_events
+        SET delivered = 't', locked = 'f'
+        WHERE id = $1
+      |] (Identity $ sefId se) True
+
+processDead :: (MonadIO m, MonadError QErr m) => Q.PGPool -> ScheduledEventFull -> m ()
```

Comment:
It seems like an event can be in various states, like “locked,” “delivered,” “cancelled,” and “dead,” but it isn’t clear to me what those things mean, which ones are mutually-exclusive, and which ones can overlap. Could you explain this in a comment somewhere? Something like `Note [Scheduled event lifecycle]` seems like a good idea!

Reply:
I have added the note explaining the different states a scheduled event can be in.